core/src/main/scala/me/wojnowski/oidc4s/IdTokenVerifier.scala

```diff
@@ -104,49 +103,74 @@ object IdTokenVerifier {
-          .toEither
-          .leftMap[Error](throwable => JwtVerificationError(throwable))
-          .flatMap { rawClaims =>
-            JsonDecoder[A]
-              .decode(rawClaims)
-              .leftMap(IdTokenVerifier.Error.CouldNotDecodeClaim.apply)
-          }
-
       private def ensureExpectedIssuer(tokenIssuer: Issuer, expectedIssuer: Issuer): Either[Error.UnexpectedIssuer, Unit] =
         Either.cond(expectedIssuer === tokenIssuer, (), IdTokenVerifier.Error.UnexpectedIssuer(tokenIssuer, expectedIssuer))
 
-      private def extractKid(headerJson: String): Either[CouldNotExtractKeyId.type, String] =
+      private def ensureNotExpired(now: Instant, expiresAt: Instant): Either[Error.TokenExpired, Unit] =
+        Either.raiseWhen(expiresAt.isBefore(now))(TokenExpired(since = expiresAt))
+
+      private def decodeHeader(headerJson: String): Either[CouldNotDecodeHeader, JwtHeader] =
         JsonDecoder[JwtHeader]
           .decode(headerJson)
-          .toOption
-          .flatMap(_.keyId)
-          .toRight(CouldNotExtractKeyId)
+          .leftMap(CouldNotDecodeHeader.apply)
+
+      private def decodeJwtAndVerifySignature[A: ClaimsDecoder](rawToken: String, key: PublicKey, header: JwtHeader)
+        : Either[Error, (A, IdTokenClaims)] =
+        rawToken.split('.') match {
+          case Array(rawHeader, rawClaims, rawSignature) =>
+            for {
+              _      <- verifyAlgorithm(header.algorithm)
+              _      <- verifySignature(header.algorithm.fullName, key, rawHeader, rawClaims, rawSignature)
+              result <- parseClaims[A](rawClaims)
+            } yield result
+
+          case _ =>
+            InvalidToken.asLeft
+        }
+
+      private def parseClaims[A: ClaimsDecoder](rawClaims: String): Either[CouldNotDecodeClaim, (A, IdTokenClaims)] =
+        Try {
+          new String(Base64.getUrlDecoder.decode(rawClaims))
+        }.toEither.leftMap(t => CouldNotDecodeClaim(t.getMessage)).flatMap { rawJson =>
+          ClaimsDecoder[A].decode(rawJson).leftMap(CouldNotDecodeClaim.apply)
+        }
+
+      private def verifyAlgorithm(algorithm: Algorithm) =
+        Either.raiseUnless(Algorithm.supportedAlgorithms.contains_(algorithm))(UnsupportedAlgorithm(algorithm.name.some))
+
+      private def verifySignature(
+        signingAlgorithm: String,
+        publicKey: PublicKey,
+        rawHeader: String,
+        rawClaims: String,
+        rawSignature: String
+      ) =
+        Try {
+          val decodedSignature = Base64.getUrlDecoder.decode(rawSignature)
+          val signatureInstance = Signature.getInstance(signingAlgorithm)
+          signatureInstance.initVerify(publicKey)
+          signatureInstance.update(s"$rawHeader.$rawClaims".getBytes(StandardCharsets.UTF_8))
+          signatureInstance.verify(decodedSignature)
+        } match {
+          case Success(true) => Either.unit
+          case _             => InvalidSignature.asLeft
+        }
```

Comment:
I wonder if wrapping in `Try` is enough for 100% purity. Perhaps it's safer to wrap it with `delay`/`blocking` in case it does nasty stuff like blocking a thread? Or maybe it's good enough, I like how this API only requires a `Monad`

Reply:
Yeah, that was a conscious decision exactly because of you noticed — dependence on `Monad` enables `Id`/`Try`/`Either` instances for those who don't use an effect system.

I agree in general it's be safer to use `delay`/`blocking`, but in this case, given these are known CPU-intensive, non-blocking tasks, I think it's a reasonable concession 🙂